sdk/search/Azure.Search.Documents/tests/Samples/FieldBuilder/FieldBuilder.cs

```diff
@@ -0,0 +1,451 @@
+                : DefaultResolver;
+            return BuildForType(modelType, namingPolicy);
+        }
+
+        /// <summary>
+        /// Creates a collection of <see cref="SearchField"/> objects corresponding to
+        /// the properties of the type supplied.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The type for which fields will be created, based on its properties.
+        /// </typeparam>
+        /// <param name="namingPolicy">
+        /// <see cref="JsonNamingPolicy"/> to use.
+        /// This ensures that the field names are generated in a way that is
+        /// consistent with the way the model will be serialized.
+        /// </param>
+        /// <returns>A collection of fields.</returns>
+        public static IList<SearchField> BuildForType<T>(JsonNamingPolicy namingPolicy) => BuildForType(typeof(T), namingPolicy);
+
+        /// <summary>
+        /// Creates a collection of <see cref="SearchField"/> objects corresponding to
+        /// the properties of the type supplied.
+        /// </summary>
+        /// <param name="modelType">
+        /// The type for which fields will be created, based on its properties.
+        /// </param>
+        /// <param name="namingPolicy">
+        /// <see cref="JsonNamingPolicy"/> to use.
+        /// Contract resolver that the SearchIndexClient will use.
+        /// This ensures that the field names are generated in a way that is
+        /// consistent with the way the model will be serialized.
+        /// </param>
+        /// <returns>A collection of fields.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="modelType"/> or <paramref name="namingPolicy"/> is null.</exception>
+        public static IList<SearchField> BuildForType(Type modelType, JsonNamingPolicy namingPolicy)
+        {
+            if (modelType is null)
+            { throw new ArgumentNullException(nameof(modelType));
+            }
+
+            if (namingPolicy is null)
+            { throw new ArgumentNullException(nameof(namingPolicy));
+            }
+
+            ArgumentException FailOnNonObjectDataType()
+            {
+                string errorMessage =
+                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " +
+                    "class or struct with public properties.";
+
+                throw new ArgumentException(errorMessage, nameof(modelType));
+            }
+
+            if (ObjectInfo.TryGet(modelType, out ObjectInfo info))
+            {
+                if (info.Properties.Length == 0)
+                {
+                    throw FailOnNonObjectDataType();
+                }
+
+                // Use Stack to avoid a dependency on ImmutableStack for now.
+                return BuildForTypeRecursive(modelType, info, namingPolicy, new Stack<Type>(new[] { modelType }));
+            }
+
+            throw FailOnNonObjectDataType();
```

Comment:
Nit - would it be simpler to restructure this as

```C#
if (ObjectInfo.TryGet(modelType, out ObjectInfo info) &&
    info.Properties.Length > 0)
{
    // Use Stack to avoid a dependency on ImmutableStack for now.
    return BuildForTypeRecursive(modelType, info, namingPolicy, new Stack<Type>(new[] { modelType }));
}

string errorMessage =
    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " +
    "class or struct with public properties.";
throw new ArgumentException(errorMessage, nameof(modelType));
```

Reply:
The final `FieldBuilder` for track 2 will be rewritten from scratch and wouldn't do this. I'm not a fan of implicit capture classes to make stuff like this work.